node/actors/bft/src/testonly/run.rs

```diff
@@ -71,77 +110,407 @@ impl Test {
+    //   don't move on from a view until they reach quorum, then "later" could be defined by so many
+    //   messages in a view that it indicates a timeout loop. NB the consensus might not actually have
+    //   an actual timeout loop and might rely on the network trying to re-broadcast forever.
+    // * OTOH if all partitions are at least quorum sized, then we there will be a subset of nodes that
+    //   can move on to the next view, in which a new partition configuration will allow them to broadcast
+    //   to previously isolated peers.
+    // * One idea is to wait until replica A wants to send to replica B in a view when they are no longer
+    //   partitioned, and then unstash all previous A-to-B messages. This would _not_ work with HotStuff
+    //   out of the box, because replicas only communicate with their leader, so if for example B missed
+    //   a LeaderCommit from A in an earlier view, B will not respond to the LeaderPrepare from C because
+    //   they can't commit the earlier block until they get a new message from A. However since
+    //   https://github.com/matter-labs/era-consensus/pull/119 the ReplicaPrepare messages are broadcasted,
+    //   so we shouldn't have to wait long for A to unstash its messages to B.
+    // * If that wouldn't be acceptable then we could have some kind of global view of stashed messages
+    //   and unstash them as soon as someone moves on to a new view.
+    let mut stashes: HashMap<Port, Vec<io::OutputMessage>> = HashMap::new();
+
+    // Either stash a message, or unstash all messages and send them to the target along with the new one.
+    let mut send_or_stash = |can_send: bool, target_port: Port, msg: io::OutputMessage| {
+        let stash = stashes.entry(target_port).or_default();
+        let view = output_msg_view_number(&msg);
+        let kind = output_msg_label(&msg);
+
+        if can_send {
+            let s = &sends[&target_port];
+
+            // Send after taking note of potentially gossipable blocks.
+            let send = |msg| {
+                if let Some(bn) = block_to_gossip(target_port, &msg) {
+                    gossip_send.send((port, target_port, bn));
+                }
+                s.send(msg);
+            };
+
+            // Messages can be delivered in arbitrary order.
+            stash.shuffle(rng);
+
+            for unstashed in stash.drain(0..) {
+                let view = output_msg_view_number(&unstashed);
+                let kind = output_msg_label(&unstashed);
+                eprintln!("   ^^^ unstashed view={view} from={port} to={target_port} kind={kind}");
+                send(unstashed);
+            }
+            eprintln!("   >>> sending view={view} from={port} to={target_port} kind={kind}");
+            send(msg);
+        } else {
+            eprintln!("   VVV stashed view={view} from={port} to={target_port} kind={kind}");
+            stash.push(msg);
+        }
```

Comment:
nit: you would get less indentation, by treating "stashing path" as early exit path, i.e:
```
if !can_send {
  stash.push(msg);
  return;
}
...
```